keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,422 @@
+        kind: Copy
+        spec:
+          path: /opt/zookeeper/server.properties
+          objectRef:
+            kind: StatefulSet # Runs on ALL pods in the statefulset
+            name: "{{ .InstanceName }}-zookeeper"
+      - name: DNS information for running pod
+        kind: Command
+        spec:
+          command: # Can be string or array
+            - nslookup
+            - google.com
+          objectRef:
+            kind: Pod
+            name: "{{ .InstanceName }}-zookeeper-0"
+    filters:
+      - name: Authentication information
+        spec:
+          regex: "^host: %w+$"
+```
+
+This key is **OPTIONAL**. Default diagnostics collection will happen regardless
+of the `diagnostics.bundle` key's presence. Note, moving to a graph-based engine
+for KUDO will make selecting of resources much easier, rather than having to
+use magical strings with templates. Future iterations of this will reduce the
+complexity of selecting resources to run commands and files on.
+
+Steps in a bundle run serially. To prevent the KUDO controller manager from
+crashing, the collector process runs in another pod as a job. **TODO**:
+Bundle collection, CRD, do we take a Velero-style approach? Where are files
+stored? Might be time to introduce a KUDO-specific Minio instance.
+
+Filtering is an important part of diagnostics collection. It enables diagnostics
+to be portably sent to third parties that should not have sensitive information
+that logs and files can contain.
+
+By default, KUDO filters all resources (and custom resources) of values
+contained within the KUDO Instance's secrets. This is configurable with the
+`diagnostics.filterSecrets` key.
+
+There may be other fields that need to be filtered. To solve for this, KUDO
+introduces the `diagnostics.bundles.filters` key in `operator.yaml`, which
+contains a list of filters that files pass through before writing to disk.
+Custom filters use either a regular expression or an object reference and
+JSONPath to derive values to filter.
+
+All filtered values appear as `**FILTERED**` in relevant logs and files.
+
+### More Notes
+
+- Do we need to introduce a notion of the collector or controller manager
+  signing and/or encrypting bundles? TBD.
+
+#### Preflight Checks
+
+## Resources
+
+### bundle.resources
+
+An individual bundle resource is represented as a list inside of the
+`diagnostics.bundle.resources` key. Resources ALWAYS have the following keys:
+
+- **name**: The human-readable name of the file.
+- **key**: The machine-readable name of the file. This is used for both
+  references (if needed in the future) and filenames. Extension is OPTIONAL,
+  but may be useful for inferring mime types.
+- **kind**: The kind of bundle item.
+- **spec**: The attributes of a particular kind. This is different for every
+  kind.
+
+Also, may specs may include an `objectRef`. It ALWAYS has the following keys:
+
+- **kind**: The Kubernetes Kind referenced. For example, this may be a
```

Comment:
Can this be a CR as well? I would assume so, but in that case do we need a GVK? Or is a GK enough?

Reply:
Hmmm.. Not sure it would make sense to allow CRs here. But in that case we need to specify what kinds we exactly allow here. Probably something that boils down to something runnable?

Reply:
I wonder if there's a convention we can follow. You could _potentially_ have a CR that is representative of a set of pods - see the metacontroller DaemonJob example. Additionally, could you just eventually reference a KUDO Instance that is created as part of an operator (see: Flink) and get ALL PODS inside of Flink?

There is some convention here around selectors and labels inside of Deployment, Statefulset, and Pod. Though, service has that too. Could you select ALL PODS that are selected by that Service? It seems weird at first, but it could be very useful (give me the traffic information for all pods that are currently being load balanced by this service).

Reply:
What's a CR?

Reply:
CR stands here for "Custom Resource"